Pitch accents

toasted-nutbread opened the pull request:
Adds support for #61.

The dictionary generation script can be found here:
https://github.com/toasted-nutbread/yomichan-pitch-accent-dictionary

A build of the dictionary can be found here:
https://github.com/toasted-nutbread/yomichan-pitch-accent-dictionary/releases/tag/1.0.0

There are three different modes for displaying pitch accent data, each of which can be enabled/disabled individually.
* **Show downstep notation for pitch accents**
  Shows a bar over characters with a higher pitch and, if present, a small down bar where the pitch drop is.
* **Show position notation for pitch accents**
  Shows the numeric indication for the accent position. E.g. [1]
* **Show graph for pitch accents**
  Shows a graph of points indicating the accent positions of the mora. Currently disabled by default.

Here are some good terms for testing:
| Term | Reason |
| ----- |:--------- |
| 端 | Single kanji term with three readings with varying accent positions. |
| に | Single term with many sequenced terms in merged mode. |
| ２つ | Term which has multiple pitch accent positions with tags. |
| 博士 | Term with two kanji which has multiple readings with the same accent position. |
| し | Term with single pitch accent definition and no sequenced terms. |
| しっ | Term with no pitch accent definition. |

Example image showing all 3 modes:
![image](https://user-images.githubusercontent.com/11037431/75633831-8f170180-5bd6-11ea-8396-d859e1a28e7c.png)

siikamiika commented:
Looks very nice, good job! I'll test this more closely some day next week.

siikamiika commented on ext/mixed/js/display-generator.js:
```diff
@@ -258,6 +276,117 @@ class DisplayGenerator {
+        DisplayGenerator._appendMultiple(n, this.createTag.bind(this), tags);
+
+        n = node.querySelector('.term-pitch-accent-expression-list');
+        DisplayGenerator._appendMultiple(n, this.createPitchExpression.bind(this), expressions);
```
Instead of `expressions`, I think this could have something similar to `only` here:
https://github.com/FooSoft/yomichan/blob/46fee07d36b3966af9bacca8c6253b044bde07ee/ext/bg/js/dictionary.js#L282-L293
Now there are issues when either multiple expressions have the same pitch accent:
![2020-03-03_00-26-34](https://user-images.githubusercontent.com/5602483/75724473-22d7f300-5ce7-11ea-93e7-c6cbc4d690fc.png)
Or only graphs are enabled:
![2020-03-03_00-27-34](https://user-images.githubusercontent.com/5602483/75724466-1f446c00-5ce7-11ea-879b-7f3aed049866.png)

toasted-nutbread replied:
> Instead of `expressions`, I think this could have something similar to `only` 

That's effectively already what it's doing

> [img1]

Is the issue here that `背、脊：` is displayed twice?

> [img2]

Is the issue here that there is no reading displayed?

siikamiika replied:
The problem is that it only works when all expressions differ in both reading and kanji. In the 背 example the pitch expression pointers are not needed because both pitch accents apply to both results, and in the 端 example the logic thinks that everything is fine although there are three different readings.

toasted-nutbread replied:
I'll look into making this implementation more consistent with the `only` implementation, but there are potentially ambiguous results with that method as well:

端/はた: Which kanji does the (はた only) apply to?
![image](https://user-images.githubusercontent.com/11037431/75942332-1aef8e80-5e60-11ea-94a6-a26b490793ea.png)

siikamiika replied:
If you mean the JMdict's definition "third person", it applies to every kanji. The kanji is left out to keep it brief. It tries to emulate the original JMdict format which looks like this (from [my abandoned project](https://github.com/siikamiika/mecab-translate)):
![2020-03-05_14-15-10](https://user-images.githubusercontent.com/5602483/75980749-c691f080-5eeb-11ea-9b11-9015fe765c36.png)

I've also been thinking of a display format where the readings are grouped under each kanji, but that could be problematic with cross browser ruby text and Anki. Better to fix the data structure first anyway https://github.com/FooSoft/yomichan/issues/337.

siikamiika commented:
Just a minor cosmetic/usability issue, but I think the titles can take quite a lot of vertical space with the default sized popup. Maybe an option to disable them would be useful.

siikamiika commented:
The pitch accents don't respect the "compact glossaries" setting. I think they look good as is and reusing the setting could be undesirable, so my solution would be to either leave it or add a separate setting. Minor issue.

siikamiika commented on ext/mixed/js/display-generator.js:
```diff
@@ -258,6 +276,117 @@ class DisplayGenerator {
+            this._termPitchAccentStaticTemplateIsSetup = true;
+            const t = this._templateHandler.instantiate('term-pitch-accent-static');
+            document.head.appendChild(t);
+        }
```
Is this worth the optimization or are there other concerns? Maybe it could be added in `prepare`.

toasted-nutbread replied:
In cases where pitch accent dictionaries aren't installed, it won't be added. I suppose it's not a big deal either way if it's unused, so it could go into prepare.

siikamiika replied:
Found this old review comment after reviewing the history, feel free to move it there now or leave as is because it's just a minor code style issue.

siikamiika commented on ext/mixed/js/display-generator.js:
```diff
@@ -346,4 +481,85 @@ class DisplayGenerator {
+    }
 }
+
+DisplayGenerator._smallKanaSet = DisplayGenerator._jpCreateSmallKanaSet();
```
There are now even more Japanese functions in `DisplayGenerator`. I remember that we let `_isCharacterKanji` slide because it was so simple, but this is more complex. Might be worth it to include `japanese.js` as a dependency now.

Also, does adding `_smallKanaSet` to the class have the same issues as callbacks (https://github.com/FooSoft/yomichan/issues/359)?

toasted-nutbread replied:
> There are now even more Japanese functions in `DisplayGenerator`. I remember that we let `_isCharacterKanji` slide because it was so simple, but this is more complex. Might be worth it to include `japanese.js` as a dependency now.

There should probably be a new japanese.js for the frontend now, yeah.

> Also, does adding `_smallKanaSet` to the class have the same issues as callbacks (#359)?

Not really an issue per se, but it could be updated to use the same convention. This change was initially made before that, so that's partially why this style was used.

siikamiika replied:
> There should probably be a new japanese.js for the frontend now, yeah.

That's cool too if there's not much overlap.

toasted-nutbread replied:
Updated in 0d80fcdf86745da133e4510eeea809a4eeafe120

seanblue commented:
I currently see all the readings for 端 as separate entries. Does this PR change that view to combine them as shown in your pictures or is that a setting? If a setting, how will this look with separate entries?

toasted-nutbread commented:
> 
> 
> I currently see all the readings for 端 as separate entries. Does this PR change that view to combine them as shown in your pictures or is that a setting? If a setting, how will this look with separate entries?

It's the "Result grouping" setting when set to "Group results by main dictionary entry".

Looks like this in the other modes:
![image](https://user-images.githubusercontent.com/11037431/75839298-89681a00-5d96-11ea-9655-8c349c2e0d4b.png)

toasted-nutbread commented:
> Just a minor cosmetic/usability issue, but I think the titles can take quite a lot of vertical space with the default sized popup. Maybe an option to disable them would be useful.

The reason I added them headings is so that the pitch accents and definitions don't blend together. Do you (or anyone else following) think it would be unclear if the headings weren't there? They were also added since there were some spacing issues with the overlines for the downstep notation, but that can probably be adjusted without headings.

siikamiika commented:
> Do you (or anyone else following) think it would be unclear if the headings weren't there?

I think the dictionary name tag gives some hint already, but I like how it currently looks on the search page where there's enough vertical space. I tested with
```css
.term-entry-body-section-header {
  display: none;
}
```
and it still seems quite clear.

seanblue commented:
> Do you (or anyone else following) think it would be unclear if the headings weren't there?

I don't think the headings are necessary. You have to add a pitch accent dictionary explicitly, so at that point I think you expect the extra information to show up. And I agree that the tags should make it clear if for some reason there was confusion.

siikamiika commented:
There's a potential gotcha in just removing the headings when you have multiple pitch accent dictionaries, pitch accents and definitions use a separate ordered list and it could look weird:
![2020-03-04_14-00-08](https://user-images.githubusercontent.com/5602483/75877483-887bca80-5e20-11ea-8ed0-1b012766c88a.png)
One solution would be to make pitch accents a subset of regular definitions, but that would be a major restructuring in the code and then you would have other problems such as sorting pitch accents before definitions. It's currently an advanced feature to sort dictionaries.

toasted-nutbread commented:
> There's a potential gotcha in just removing the headings when you have multiple pitch accent dictionaries, pitch accents and definitions use a separate ordered list and it could look weird:

It may also be possible to just show the index for both at all times, and offset the starting index of the standard dictionary entries. Not sure if this would be confusing though; could do something like change the color of the dictionary tag to more clearly indicate it's for a separate category.

siikamiika commented:
> It may also be possible to just show the index for both at all times, and offset the starting index of the standard dictionary entries. Not sure if this would be confusing though; could do something like change the color of the dictionary tag to more clearly indicate it's for a separate category.

Maybe it's best to keep it as simple as possible, this could also require significant changes to the Anki templates. I think a separator with separate indices and possibly different dictionary tag color are fine.

Speaking of Anki, it doesn't seem to be possible to export pitch accents to Anki right now. Do you know if it's possible to render SVG in the browser as PNG easily and how the downstep notation styles can be exported?

toasted-nutbread commented:
> Speaking of Anki, it doesn't seem to be possible to export pitch accents to Anki right now. Do you know if it's possible to render SVG in the browser as PNG easily

It's probably possible to export as a PNG using a `<canvas>` to render, and I think Anki also just supports using `<svg>` tags.

> and how the downstep notation styles can be exported?

Would probably have to just be different `{tags}` for the possible styles.

Anki support should probably be part of a follow-up PR so as to limit the scope of this PR.

siikamiika commented:
> > Speaking of Anki, it doesn't seem to be possible to export pitch accents to Anki right now. Do you know if it's possible to render SVG in the browser as PNG easily
> 
> It's probably possible to export as a PNG using a `<canvas>` to render, and I think Anki also just supports using `<svg>` tags.
> 
> > and how the downstep notation styles can be exported?
> 
> Would probably have to just be different `{tags}` for the possible styles.

Nice :+1: 

> Anki support should probably be part of a follow-up PR so as to limit the scope of this PR.

Sure, then I think the only issue is removing the headings and deciding how to handle indexing. Right now a horizontal bar separator seems like the best option to me.

toasted-nutbread commented:
Styles have been updated and the disambiguation notation is now more consistent with the existing "only" implementation.

siikamiika commented on ext/mixed/js/display-generator.js:
```diff
@@ -498,10 +499,22 @@ class DisplayGenerator {
+                if (!areSetsEqual(resultExpressions, allExpressions)) {
+                    exclusiveExpressions.push(...getSetIntersection(resultExpressions, allExpressions));
+                }
+                result.exclusiveExpressions = exclusiveExpressions;
```
I'd update it like this to handle readings as well:
```diff
diff --git a/ext/mixed/js/display-generator.js b/ext/mixed/js/display-generator.js
index 8177c31c..b6469d62 100644
--- a/ext/mixed/js/display-generator.js
+++ b/ext/mixed/js/display-generator.js
@@ -482,10 +482,13 @@ class DisplayGenerator {
         const results = new Map();

         const allExpressions = new Set();
+        const allReadings = new Set();
         const expressions = definition.expressions;
         const sources = Array.isArray(expressions) ? expressions : [definition];
         for (const {pitches: expressionPitches, expression} of sources) {
+            allExpressions.add(expression);
             for (const {reading, pitches, dictionary} of expressionPitches) {
+                allReadings.add(reading);
                 let dictionaryResults = results.get(dictionary);
                 if (typeof dictionaryResults === 'undefined') {
                     dictionaryResults = [];
@@ -499,7 +502,6 @@ class DisplayGenerator {
                         dictionaryResults.push(pitchInfo);
                     }
                     pitchInfo.expressions.add(expression);
-                    allExpressions.add(expression);
                 }
             }
         }
@@ -508,9 +510,13 @@ class DisplayGenerator {
             for (const result of dictionaryResults) {
                 const exclusiveExpressions = [];
                 const resultExpressions = result.expressions;
+                const resultReadings = new Set([result.reading]);
                 if (!areSetsEqual(resultExpressions, allExpressions)) {
                     exclusiveExpressions.push(...getSetIntersection(resultExpressions, allExpressions));
                 }
+                if (!areSetsEqual(resultReadings, allReadings)) {
+                    exclusiveExpressions.push(...getSetIntersection(resultReadings, allReadings));
+                }
                 result.exclusiveExpressions = exclusiveExpressions;
             }
         }
```

toasted-nutbread replied:
> I'd update it like this to handle readings as well:

The readings are redundant when the downstep notation is also visible. Does it make sense to not display them for this case? (I think this is why I initially didn't add this, because I forgot that there were options to turn that off.)

![image](https://user-images.githubusercontent.com/11037431/77851884-8c91d280-71a9-11ea-8f6d-ac892dcd118c.png)

siikamiika replied:
Yeah, I think it's good to add that exception.

toasted-nutbread replied:
Updated.
